server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -655,9 +659,117 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
       );
     }
 
+    if (config.isKillSegments()) {
+      helpers.add(
+          new DruidCoordinatorSegmentKiller(
+              metadataSegmentManager,
+              indexingServiceClient,
+              config.getCoordinatorKillDurationToRetain(),
+              config.getCoordinatorKillPeriod(),
+              config.getCoordinatorKillMaxSegments()
+          )
+      );
+    }
+
     return ImmutableList.copyOf(helpers);
   }
 
+  static class DruidCoordinatorSegmentKiller implements DruidCoordinatorHelper
+  {
+    
+    private final long period;
+    private final long retainDuration;
+    private final int maxSegmentsToKill;
+    private long lastKillTime = 0;
+
+
+    private final MetadataSegmentManager segmentManager;
+    private final IndexingServiceClient indexingServiceClient;
+
+    DruidCoordinatorSegmentKiller(
+        MetadataSegmentManager segmentManager,
+        IndexingServiceClient indexingServiceClient,
+        Duration retainDuration,
+        Duration period,
+        int maxSegmentsToKill
+    )
+    {
+      this.period = period.getMillis();
+      Preconditions.checkArgument(this.period > 0, "coordinator kill period must be > 0");
+
+      this.retainDuration = retainDuration.getMillis();
+      Preconditions.checkArgument(this.retainDuration >= 0, "coordinator kill retainDuration must be >= 0");
+
+      this.maxSegmentsToKill = maxSegmentsToKill;
+      Preconditions.checkArgument(this.maxSegmentsToKill > 0, "coordinator kill maxSegments must be > 0");
+
+      log.info(
+          "Kill Task scheduling enabled with period [%s], retainDuration [%s], maxSegmentsToKill [%s]",
+          this.period,
+          this.retainDuration,
+          this.maxSegmentsToKill
+      );
+
+      this.segmentManager = segmentManager;
+      this.indexingServiceClient = indexingServiceClient;
+    }
+
+    @Override
+    public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+    {
+      Set<String> whitelist = params.getCoordinatorDynamicConfig().getKillDataSourceWhitelist();
+
+      if (whitelist != null && whitelist.size() > 0 && (lastKillTime + period) < System.currentTimeMillis()) {
+        lastKillTime = System.currentTimeMillis();
+
+        for (String dataSource : whitelist) {
+          final Interval intervalToKill = findIntervalForKillTask(dataSource, maxSegmentsToKill);
+          if (intervalToKill != null) {
+            try {
+              indexingServiceClient.killSegments(dataSource, intervalToKill);
```

Comment:
not really, in the worst case it might run kill-task one more time right after someone updated the whitelist. (which can happen otherwise too just with lower probability).